src/Bicep.Core/Emit/TemplateWriter.cs

```diff
@@ -85,35 +91,45 @@ private static string GetSchema(ResourceScope targetScope)
 
             return "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#";
         }
+
         private readonly EmitterContext context;
         private readonly EmitterSettings settings;
+        private readonly IDictionary<string, IDictionary<int, IList<(int start, int end)>>> rawSourceMap;
+
+        public Dictionary<int, (string, int)>? SourceMap; // ARM JSON line => (Bicep File, Bicep Line)
 
         public TemplateWriter(SemanticModel semanticModel, EmitterSettings settings)
         {
             this.context = new EmitterContext(semanticModel, settings);
             this.settings = settings;
+            this.rawSourceMap = new Dictionary<string, IDictionary<int, IList<(int, int)>>>();
         }
 
         public void Write(JsonTextWriter writer)
         {
-            // Template is used for calcualting template hash, template jtoken is used for writing to file.
+            // Template is used for calculating template hash, template JToken is used for writing to file.
             var (template, templateJToken) = GenerateTemplateWithoutHash();
+
             var templateHash = TemplateHelpers.ComputeTemplateHash(template.ToJToken());
             if (templateJToken.SelectToken(GeneratorMetadataPath) is not JObject generatorObject)
             {
                 throw new InvalidOperationException($"generated template doesn't contain a generator object at the path {GeneratorMetadataPath}");
             }
-            generatorObject.Add("templateHash", templateHash);
+            generatorObject.Add(TemplateHashPropertyName, templateHash);
```

Comment:
I'm wondering if it's worth changing the way this works to make the calculations later on easier. Right now we emit without the hash and then add the property in on a second pass, which obviously changes all the calculations.

We could consider emitting a placeholder the same length as the hash, and then just replacing it a second pass - to avoid changing any positions - e.g.:
```
"templateHash": "000000000000",
```
-> 
```
"templateHash": "002194837493",
```

Reply:
So I am not sure how I would account for the hash padding ahead of time before the template is output? I tried this but as expected the hash length varies which messes up the source map. Maybe I could prefix zeroes to shorter actual hashes to make all hashes a set length?

Otherwise, the current approach seems like the simplest solution.